### ISSUE TITLE
Ubuntu 23.04 Support

### ISSUE THREAD

**Issue by Bpoole908** [NVIDIA/nvidia-container-toolkit#72]
Are there any plans to support Ubuntu 23.04?

**Comment by elezar** [comment]
@Bpoole908 we generally only publish packages for LTS releases. With that said, the Ubuntu18.04 repositories can be used for all newer Ubuntu distributions (and are already used for the 20.04 and 22.04 packages).

This means following the standard instructions [here](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#installing-on-ubuntu-and-debian), but explicitly setting `distribution=ubuntu18.04`:

```
distribution=ubuntu18.04 \
      && curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
      && curl -s -L https://nvidia.github.io/libnvidia-container/$distribution/libnvidia-container.list | \
            sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
            sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
```

**Comment by jawadhaidar** [comment]
I have ubuntu 23.04 as well and tried this modification but still got the same error.  @CallMeAl1973 did it work on your ubuntu ? thank you in advance 

docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].
ERRO[0004] error waiting for container: context canceled

**Comment by elezar** [comment]
@jawadhaidar please see our updated instructions for installing the NVIDIA Container Toolkit. This includes instructions to configure Docker. 

https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html#installing-with-apt

**Comment by jawadhaidar** [comment]
@elezar thank you for support. I tried the exact steps (installation + configuration) on two workstations (both having gpus) with ubuntu 22.04 and ubuntu 23.04 respectively. However, it only worked on the one having ubuntu 22.04. 
for the workstation having ubuntu 23.04 I tried : "docker run --gpus all -it tensorflow/tensorflow:latest-gpu" and got the following error:  
"docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].
ERRO[0000] error waiting for container: context canceled "

**Comment by elezar** [comment]
@jawadhaidar how is docker installed on the two machines?

**Comment by jawadhaidar** [comment]
Both using the same instructions "[Install using the apt repository]" found on the official website of docker 
 https://docs.docker.com/engine/install/ubuntu/

**Comment by elezar** [comment]
@jawadhaidar could you try run your examples specifying the nvidia runtime specifically. This would mean adding --runtime nvidia to the docker run command line.

It also seems as if the issue might be that nvidia-persistenced is not running on the 23.04 machine. Could you confirm this by checking the nvidia-smi output for both machines?

**Comment by jawadhaidar** [comment]
nvidia-smi on ubuntu 23.04 result is attached in the image. 
<img width="405" alt="nvidia-smi-on23" src="https://github.com/NVIDIA/nvidia-container-toolkit/assets/74460048/b6545c21-e35d-4007-a338-e8c070f3b5ec">

command1: docker run --runtime nvidia --gpus all -it tensorflow/tensorflow:latest-gpu
returne1: docker: Error response from daemon: unknown or invalid runtime name: nvidia.

however,
command2: sudo docker run --rm --runtime=nvidia --gpus all ubuntu nvidia-smi
return2: attached image "runtimenvidia"
<img width="471" alt="runtimenvidia" src="https://github.com/NVIDIA/nvidia-container-toolkit/assets/74460048/261ee2ce-64b8-4139-9057-3eb928883d32">

**Comment by jawadhaidar** [comment]
and nvidia smi on ubunto 22.04 gives:

<img width="289" alt="nvidia-smi-on22" src="https://github.com/NVIDIA/nvidia-container-toolkit/assets/74460048/da384f2c-f8c9-4912-8e03-0e2c3323c367">

**Comment by jynrock** [comment]
I have the same issue on ubuntu 23.10:
docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].
ERRO[0006] error waiting for container: context canceled

Then I have to downgrade to ubuntu 22.04 ?

**Comment by jawadhaidar** [comment]
@jynrock I tried several things on ubuntu 23.04 and it did not work so I downgraded to 22.04 and it worked. But one thing to try is the following: in case you downloaded the docker desktop as well as docker cli the context might be causing the problem.  docker desktop and docker cli will have two different contexts so the images are not common. Additionally, docker desktop might have access to the gpu. So make sure your image is found in the default context and then try the command.

**Comment by jynrock** [comment]
Yes, I've read it wrong and I can't follow it through to the end.
Ubuntu 23.10 is not compatible with Nvidia-container-toolkit :
https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/supported-platforms.html

**Comment by ArangoGutierrez** [comment]
E2E of the toolkit currently works on the latest Ubuntu. 

I am closing this issue. If any problems persist using the latest version of the NVIDIA Container Toolkit, please open a new issue.

**Comment by elezar** [comment]
Assuming that the CUDA drivers are available on the host, our packages should be compatible with most ubuntu versions after 18.04. Please follow the distro-agnostic instructions here https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html#with-apt-ubuntu-debian and be sure to configure Docker to be aware of the `nvidia` runtime.
